### PR TITLE
Drn 252 latest listing pages fix signed again

### PR DESCRIPTION
**Pull request by mattrobertsky**

### CODE REVIEW

File: app/uk/gov/hmrc/disguisedremunerationfrontend/views/uniform/standard_field.scala.html [hunk 1]
```diff
@@ -17,13 +17,26 @@
 @import cats.implicits._
 @import ltbs.uniform._
 @import uk.gov.hmrc.disguisedremunerationfrontend.views.html.helpers._
+@import scala.util.Try
 
-@(keyList: List[String], errors: ErrorTree, fieldType: String, messages: UniformMessages[Html])(inner: Html)
+@(keyListDirty: List[String], errors: ErrorTree, fieldType: String, messages: UniformMessages[Html])(inner: Html)
+
+@keyList = @{keyListDirty match {
```
Comment (adamdye-hmrc) [inline app/uk/gov/hmrc/disguisedremunerationfrontend/views/uniform/standard_field.scala.html:24]: Is all of this still required?

File: conf/app.routes [hunk 2]
```diff
@@ -20,15 +20,21 @@ GET        /contact-details/:id                     uk.gov.hmrc.disguisedremuner
...
-POST       /scheme/new/                             uk.gov.hmrc.disguisedremunerationfrontend.controllers.JourneyController.addScheme(id = "")
-POST       /scheme/new/:id                          uk.gov.hmrc.disguisedremunerationfrontend.controllers.JourneyController.addScheme(id: String)
-
-GET        /scheme/:schemeIndex/                    uk.gov.hmrc.disguisedremunerationfrontend.controllers.JourneyController.editScheme(schemeIndex: Int, id = "")
-GET        /scheme/:schemeIndex/:id                 uk.gov.hmrc.disguisedremunerationfrontend.controllers.JourneyController.editScheme(schemeIndex: Int, id: String)
-POST       /scheme/:schemeIndex/                    uk.gov.hmrc.disguisedremunerationfrontend.controllers.JourneyController.editScheme(schemeIndex: Int, id = "")
-POST       /scheme/:schemeIndex/:id                 uk.gov.hmrc.disguisedremunerationfrontend.controllers.JourneyController.editScheme(schemeIndex: Int, id: String)
+
+#GET         /scheme-listing/                        uk.gov.hmrc.disguisedremunerationfrontend.controllers.JourneyController.schemeListAction(id = "")
```
Comment (adamdye-hmrc) [inline conf/app.routes:24]: Do you want to remove these if they are not in use?

File: conf/messages.cy [hunk 3]
```diff
@@ -481,4 +481,37 @@ accessibility.statement.howtested.p1=Cafodd y gwasanaeth ei brofi ddiwethaf ar 7
 accessibility.statement.howtested.p2.intro=Adeiladwyd y gwasanaeth gan ddefnyddio rhannau a brofwyd gan y
 accessibility.statement.howtested.p2.link=Ganolfan Hygyrchedd Digidol (Digital Accessibility Centre)
 accessibility.statement.howtested.p2.outro=Profwyd y gwasanaeth llawn gan CThEM, ac roedd y gwaith o brofi’r gwasanaeth yn cynnwys defnyddwyr anabl.
-accessibility.statement.howtested.p3=Paratowyd y dudalen hon ar 23 Medi 2019. Cafodd ei diweddaru ddiwethaf ar 23 Medi 2019.
\ No newline at end of file
+accessibility.statement.howtested.p3=Paratowyd y dudalen hon ar 23 Medi 2019. Cafodd ei diweddaru ddiwethaf ar 23 Medi 2019.
+
+schemes.heading=Schemes TODO
+schemes.title=Rydych wedi ychwanegu {0} cynllun
```
Comment (adamdye-hmrc) [inline conf/messages.cy:487]: A few TODO messages in english here

File: conf/messages.en [hunk 4]
```diff
@@ -499,3 +501,41 @@ accessibility.statement.howtested.heading=How we tested this service
 accessibility.statement.howtested.p1=The service was last tested on 5 September 2019 and was checked for compliance with WCAG 2.1 AA.
 accessibility.statement.howtested.p2=The full service was tested by HMRC and included disabled users.
 accessibility.statement.howtested.p3=This page was prepared on 23 September 2019. It was last updated on 23 September 2019.
+
+# TODO Welsh
```
Comment (adamdye-hmrc) [inline conf/messages.en:505]: And this one

File: test/uk/gov/hmrc/disguisedremunerationfrontend/data/JsonEncodingPropertySpec.scala [hunk 5]
```diff
@@ -19,20 +19,32 @@ package uk.gov.hmrc.disguisedremunerationfrontend.data
...
   "State" should s"be equal after serializing and deserializing ${timesToTest.value} times" in {
     forAll(StateGen.genJourneyState, minSuccessful(timesToTest)) { state =>
       Json.fromJson[JourneyState](Json.toJson(state)) shouldBe JsSuccess(state)
     }
   }
 
+  "DB" should s"be equal after serializing and deserializing ${timesToTest.value} times" in {
+    pending
+    forAll(DBGen.genDB, minSuccessful(timesToTest)) { db =>
+      println(s"db is $db")
```
Comment (adamdye-hmrc) [inline test/uk/gov/hmrc/disguisedremunerationfrontend/data/JsonEncodingPropertySpec.scala:43]: Do you want to println this every time?

File: app/uk/gov/hmrc/disguisedremunerationfrontend/aa_data/AboutYou.scala [hunk 6]
```diff
@@ -76,28 +76,28 @@ object AboutYou {
         for {
           id <- ask[Either[Nino, Utr]]("about-scheme-user",
             validation = List(
-              Rule.condAtPath[Either[Nino,Utr]]("Left", "a")(
+              Rule.condAtPath[Either[Nino,Utr]]("Right", "b")(
```
[the remark this answers is not in the document]
Comment (mattrobertsky) [reply app/uk/gov/hmrc/disguisedremunerationfrontend/aa_data/AboutYou.scala:79]: I can't remember but there would have been a reason

File: app/uk/gov/hmrc/disguisedremunerationfrontend/controllers/JourneyController.scala [hunk 7]
```diff
@@ -391,7 +398,7 @@ class JourneyController @Inject()(
   )(
     implicit request: AuthorisedRequest[AnyContent]
   ): Html = {
-    personalDetailsBlock(state, retrievedNino) |+| schemeBlock(state)
+    Html(personalDetailsBlock(state, retrievedNino).toString + schemeBlock(state).toString)
```
[the remark this answers is not in the document]
Comment (mattrobertsky) [reply app/uk/gov/hmrc/disguisedremunerationfrontend/controllers/JourneyController.scala:401]: again I can't remember - maybe something to do with HTML concatenation

File: app/uk/gov/hmrc/disguisedremunerationfrontend/repo/Save4LaterPersistence.scala [hunk 8]
```diff
@@ -38,23 +38,46 @@ trait Persistence {
...
-      mapFormatter.reads(json).map{ _.map { case (k,v) => (k.split("/").toList, v)} }
-  }
+  implicit val dbFormatter: Format[DB] = DbFormat.dbFormatter
 
   def makeKey(path: String): String = {
     path.replaceFirst("^.*\\/disguised-remuneration\\/", "").split("/").toList match {
-      case a if a.exists(_.matches("[0-9]")) => a.slice(0, a.lastIndexWhere(_.matches("[0-9]")) +1).mkString("/")
-      case b if b.exists(_.matches("new")) => b.slice(0, b.lastIndexWhere(_.matches("new")) +1).mkString("/")
-      case c => c.head
+      case foo if foo.exists(_.matches("schemes")) =>
```
[the remark this answers is not in the document]
Comment (mattrobertsky) [reply app/uk/gov/hmrc/disguisedremunerationfrontend/repo/Save4LaterPersistence.scala:73]: foo is a valid variable name - and I needed to go ahead of a,b,c or do a load of renaming, also I considered the approach to be something of a necessary hack, the use of foo implies this

File: app/uk/gov/hmrc/disguisedremunerationfrontend/repo/Save4LaterPersistence.scala [hunk 9]
```diff
@@ -81,26 +104,30 @@ abstract class Save4LaterPersistence @Inject() (
...
+            } else {
+              (b \ cacheRepositoryKey).toOption
+            }
+          }
+          d <- OptionT.fromOption[Future]{
+            (b \ path \ "scheme-name").isDefined.some
+          }
+        } yield (c.as[DB],d)}.fold(Map.empty[List[String],String]) {
+          case (a, true) => a.map(x => ("schemes" :: "add" :: x._1, x._2))
+          case b => b._1
```
[the remark this answers is not in the document]
Comment (mattrobertsky) [reply app/uk/gov/hmrc/disguisedremunerationfrontend/repo/Save4LaterPersistence.scala:124]: NA

**Pull request merged**